server/src/main/java/org/elasticsearch/common/LocalTimeOffset.java

```diff
@@ -0,0 +1,679 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.zone.ZoneOffsetTransition;
+import java.time.zone.ZoneOffsetTransitionRule;
+import java.time.zone.ZoneRules;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Converts utc into local time and back again.
+ * <p>
+ * "Local time" is defined by some time zone, specifically and {@link ZoneId}.
+ * At any point in time a particular time zone is at some offset from from
+ * utc. So converting from utc is as simple as adding the offset.
+ * <p>
+ * Getting from local time back to utc is harder. Most local times happen once.
+ * But some local times happen twice. And some don't happen at all. Take, for
+ * example, the time in my house. Most days I don't touch my clocks and I'm a
+ * constant offset from UTC. But once in the fall at 2am I roll my clock back.
+ * So at 5am utc my clocks say 1am. Then at 6am utc my clocks say 1am AGAIN.
+ * I do similarly terrifying things again in the spring when I skip my clocks
+ * straight from 1:59am to 3am.
+ * <p>
+ * So there are three methods to convert from local time back to utc,
+ * {@link #localToUtc(long, Strategy)}, {@link #localToSensibleUtc(long)} and
+ * {@link #localToUtcInThisOffset(long)}. They are all unique and frightening.
+ */
+public abstract class LocalTimeOffset {
+    /**
+     * Lookup offsets for a provided zone. This <strong>can</strong> fail if
+     * there are many transitions and the provided lookup would be very large.
+     *
+     * @return a {@linkplain Lookup} or {@code null} if none could be built 
+     */
+    public static Lookup lookup(ZoneId zone, long minUtcMillis, long maxUtcMillis) {
+        if (minUtcMillis > maxUtcMillis) {
+            throw new IllegalArgumentException("[" + minUtcMillis + "] must be <= [" + maxUtcMillis + "]");
+        }
+        ZoneRules rules = zone.getRules();
+        {
+            LocalTimeOffset fixed = checkForFixedZone(zone, rules);
+            if (fixed != null) {
+                return new FixedLookup(zone, fixed);
```

Comment:
Do you need to special-case this here? We handle the more general case that the zone is fixed in the given range in the caller.

Reply:
I'm not sure! I added that when I first wrote the class because it felt nice to get the simple stuff out of the way but I'll see!

Reply:
Thinking about caching - if we end up applying caching I'd want to apply it below this. In other words, the special case is useful for caching because we don't want to try an cache fixed zones.